File: 修订补充示例/第二轮/ch09_正文代码框.R
# 正文代码框顺序运行；扩展数据凡带“模拟”均非真实调查记录。
args <- commandArgs(trailingOnly=FALSE)
file_arg <- args[grepl("^--file=",args)]
script_dir <- if(length(file_arg)) dirname(normalizePath(gsub("~+~"," ",sub("^--file=","",file_arg[1]),fixed=TRUE))) else getwd()
repo_dir <- normalizePath(file.path(script_dir,"..",".."))
local_lib <- file.path(repo_dir,"..",".R-library")
if(dir.exists(local_lib)) .libPaths(c(local_lib,.libPaths()))
setwd(repo_dir)
library(sandwich)
library(lmtest)
set.seed(922)
acs <- read.csv(file.path(repo_dir,"Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_acs_high_income_analysis_data.csv"),fileEncoding="UTF-8-BOM")
credit <- read.csv(file.path(repo_dir,"Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_credit_default_analysis_data.csv"),fileEncoding="UTF-8-BOM")
choice_df <- read.csv(file.path(repo_dir,"修订补充示例/第二轮/ch09_有序多项选择_模拟.csv"),fileEncoding="UTF-8-BOM")

# 代码框 1: {R 中估计线性概率模型}
data_path <- file.path("Chapter09_Binary_dependent_variable_ch_logit",
                       "results", "chapter09_credit_default_analysis_data.csv")
credit <- read.csv(data_path, fileEncoding="UTF-8-BOM")
lpm <- lm(default ~ pay_delay + male + age + limit_bal_10k,
          data = credit)
lmtest::coeftest(lpm, vcov. = sandwich::vcovHC(lpm, type="HC1"),
                 df = Inf)
range(predict(lpm))


# 代码框 2: {R 中 logit 模型}
acs <- read.csv("Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_acs_high_income_analysis_data.csv", fileEncoding="UTF-8-BOM")
mod <- glm(high_income ~ bachelor + log(age) + female + employed,
           data = acs, family = binomial(link = "logit"))
exp(coef(mod))


# 代码框 3: {R 中计算平均边际效应}
mod_credit <- glm(default ~ pay_delay + male + age + limit_bal_10k,
                  data = credit, family = binomial(link = "logit"))
p0 <- predict(mod_credit, type = "response")
credit_hi <- credit
credit_hi$pay_delay <- credit_hi$pay_delay + sd(credit$pay_delay)
p1 <- predict(mod_credit, newdata = credit_hi, type = "response")
mean(p1 - p0)


# 代码框 4: {R 中从概率到分类}
phat <- predict(mod_credit, type = "response")
class_05 <- as.integer(phat >= 0.5)
table(predicted = class_05, observed = credit$default)
class_03 <- as.integer(phat >= 0.3)
table(predicted = class_03, observed = credit$default)


# 代码框 5: {R 中有序 logit 与多项 logit}
library(MASS)
choice_df <- read.csv("修订补充示例/第二轮/ch09_有序多项选择_模拟.csv")
# 有序 logit（比例赔率模型）
choice_df$edu_level <- ordered(choice_df$edu_level,
                               levels = c("低", "中", "高"))
fit_ord <- polr(edu_level ~ log_income + age, data = choice_df,
               method = "logistic", Hess = TRUE)
# 多项 logit
library(nnet)
fit_mnl <- multinom(transport ~ income + distance, data = choice_df)

File: 修订补充示例/第二轮/ch09_正文代码框.py
# 正文代码框顺序运行；带“模拟”的扩展数据不是真实调查记录。
from pathlib import Path
import numpy as np,pandas as pd,statsmodels.formula.api as smf
from scipy import stats
repo_dir=Path(__file__).resolve().parents[2]
import os
os.chdir(repo_dir)
acs=pd.read_csv(repo_dir / 'Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_acs_high_income_analysis_data.csv')
credit=pd.read_csv(repo_dir / 'Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_credit_default_analysis_data.csv')
choice_df=pd.read_csv(repo_dir / '修订补充示例/第二轮/ch09_有序多项选择_模拟.csv')

# 代码框 1: {Python 中估计线性概率模型}
import numpy as np, pandas as pd
import statsmodels.formula.api as smf
from pathlib import Path
data_path = (Path("Chapter09_Binary_dependent_variable_ch_logit")
             / "results" / "chapter09_credit_default_analysis_data.csv")
credit = pd.read_csv(data_path)
lpm = smf.ols(
    "default ~ pay_delay + male + age + limit_bal_10k",
    data=credit
).fit(cov_type="HC1", use_t=False)
credit["pred_lpm"] = lpm.predict(credit)
credit["pred_lpm"].agg(["min", "max"])


# 代码框 2: {Python 中 logit 模型}
acs = pd.read_csv("Chapter09_Binary_dependent_variable_ch_logit/results/chapter09_acs_high_income_analysis_data.csv")
mod = smf.logit(
    "high_income ~ bachelor + np.log(age) + female + employed",
    data=acs
).fit()
np.exp(mod.params)


# 代码框 3: {Python 中计算平均风险差}
import statsmodels.formula.api as smf

mod_credit = smf.logit(
    "default ~ pay_delay + male + age + limit_bal_10k",
    data=credit
).fit()
p0 = mod_credit.predict(credit)
credit_hi = credit.copy()
credit_hi["pay_delay"] += credit["pay_delay"].std()
p1 = mod_credit.predict(credit_hi)
(p1 - p0).mean()


# 代码框 4: {Python 中比较分类阈值}
phat = mod_credit.predict(credit)
for cutoff in [0.3, 0.5]:
    yhat = (phat >= cutoff).astype(int)
    print(pd.crosstab(yhat, credit["default"],
                      rownames=["预测"], colnames=["实际"]))


# 代码框 5: {Python 中有序 logit 与多项 logit}
choice_df = pd.read_csv("修订补充示例/第二轮/ch09_有序多项选择_模拟.csv")
import statsmodels.api as sm
from statsmodels.miscmodels.ordinal_model import OrderedModel
choice_df["edu_level"] = pd.Categorical(
    choice_df["edu_level"], categories=["低", "中", "高"], ordered=True)
res_ord = OrderedModel(choice_df["edu_level"],
    choice_df[["log_income", "age"]], distr="logit").fit(method="bfgs")
transport = pd.Categorical(choice_df["transport"])
X_mnl = sm.add_constant(choice_df[["income", "distance"]])
res_mnl = sm.MNLogit(transport.codes, X_mnl).fit()
print(res_mnl.summary())

